semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,148 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+open AST_generic
+module MV = Metavariable
+
+let logger = Logging.get_logger [ __MODULE__ ]
+
+(******************************************************************************)
+(* Handles AST printing for the purposes of autofix.
+ *
+ * The main printing logic happens elsewhere. This module's main purpose is to
+ * extend the existing AST printers so that they can avoid printing AST nodes
+ * which are lifted unchanged from either the target file (via metavariable
+ * bindings) or the fix pattern in the rule. This serves two purposes:
+ * - It lets us synthesize autofixes without having to implement printing for
+ *   all of the AST nodes in the fixed AST.
+ * - It lets us make minimal changes to target files by carrying over
+ *   formatting, comments, etc. from the original code.
+ *)
+(******************************************************************************)
+
+type ast_node_source =
+  (* Indicates that a node came from the target file via a metavar binding *)
+  | Target
+  (* Indicates that a node came from the fix pattern *)
+  | FixPattern
+
+(* So that we can print by lifting the original source for unchanged AST nodes,
+ * this indicates whether a given AST node came from the target file (via
+ * metavariable bindings) or from the rule's fix pattern. Nodes that do not
+ * appear in this table came from neither. *)
+type ast_node_table = (AST_generic.any, ast_node_source) Hashtbl.t
+
+module PythonPrinter = Hybrid_print.Make (struct
+  class printer = Ugly_print_AST.python_printer
+end)
+
+let get_printer lang external_printer : Ugly_print_AST.printer_t option =
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ ->
+      logger#info "Failed to render autofix: no printer available for %s"
+        (Lang.to_string lang);
+      None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
+  let starti = start.Parse_info.charpos in
+  let _, _, endi = Parse_info.get_token_end_info end_ in
+  let len = endi - starti in
+  let str = String.sub source starti len in
+  Some str
+
+(* Add each metavariable value to the lookup table so that it can be identified
```

Comment:
I considered this, but I'm reluctant to add a field to basically every AST node to support an operation that occurs on only a very small minority of AST nodes. It would significantly increase memory usage throughout Semgrep, I expect.

We'd also need to then make sure that we set that origin to `None` or `Synthesized` or something when the mapper changes a node, because otherwise we'd incorrectly pull text from the pattern or target when in reality the node has changed and we need to print it ourselves.